src/interleave-generator.mjs

```diff
@@ -0,0 +1,42 @@
+import { ensureIterable } from './internal/iterable'
+import map from './map'
+
+class InterleaveBuffer {
+  constructor (iterator) {
+    this._iterator = iterator
+    this._next = this._iterator.next()
+  }
+
+  take () {
+    const next = this._next
+    if (next.done) {
+      return undefined
+    }
+
+    this._next = this._iterator.next()
+    return next.value
+  }
+
+  canTake () {
+    return this._next.done ? null : this
```

Comment:
Wait, I thought `buffer.canTake()` returns `boolean`? Your TS declaration also says it returns `boolean`!

Reply:
Ah. Yeah I need to update the types and docs. There's a couple reasons it makes sense to do this instead of boolean. In the async variant it saves some really gross code. It also permits doing clever things like `buffer.canTake()?.take()`. You can only do that though if the type system tells you what the type really is.

Reply:
I think `boolean` is more straightforward. As for the async variant, why don't you just return a promise of (sync) `InterleaveBuffer` (`Promise<InterleaveBuffer<T>>`) instead of making an `AsyncInterleaveBuffer<T>` where each method returns a promise?

Reply:
That won't work because the buffers are passed in as arguments. We'd pass a different promise for every item, which means we'd have to pass a function `getBufferPromise()` which I think is not cleaner than what the code does now.

Reply:
Okay, so you want `buffer.canTake()?.take()` which is just a slightly shorter version of `buffer.canTake() && buffer.take()`. I don't this is compelling enough of a reason to abandon TS checker.

Your async argument is a good point though, but I wouldn't do a `(await buffer.canTake())?.take()` for it looks pretty cluttered.

Beside, user doesn't want `undefined` in their resulting iterable for most of the times, so they would still use `if` anyway.